### PR TITLE
Add a new endpoint for dev subcommand

### PR DESCRIPTION
**Pull request by hhzguo**
https://github.com/bd-iaas-us/AILint/issues/45

Add an API for dev,  github issue and prompt supported.
   """
    API for handle "dev" sub command
    Currently only the following tasks supported:
        1. github issue:
        {
            "issue": {
                "owner": "bd-iaas-us",
                "repo": "AILint",
                "issue_number": "46"
            }
        }

        2. prompt
        {"prompt": "Description of your task"}

    Each request can have only one of the above
    """

### CODE REVIEW

File: server/april/issue.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    The worker function to handle a dev prompt
+    prompt: The description of the task 
+    """
+    if prompt is None:
+        return JSONResponse(status_code=400, content={'message': "prompt cannot be empty"})
+     
+    logger.info(f'processing prompt: {prompt}')
+ 
+    agentResp = agent.call(prompt)
+    return JSONResponse(status_code=501, content={'message': "SWEAgent not connected yet"})
```
Comment (thesues) [inline server/april/issue.py:85]: so always return 400 or 501?

File: server/april/issue.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
+
+import os
+import sys
+import requests
+import json
+from fastapi.responses import JSONResponse, Response, StreamingResponse
+
+# Require pygithub installed
```
Comment (thesues) [inline server/april/issue.py:8]: do not need this line

File: server/april/issue.py (hunk 1 above)
Comment (hhzguo) [reply server/april/issue.py]: Not connected to swe agent yet

Comment (hhzguo) [reply server/april/issue.py]: There is another issue for swe agent integration

File: server/april/issue.py (hunk 2 above)
Comment (hhzguo) [reply server/april/issue.py]: Removed

File: server/april/issue.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+            if index < task.get_last_history_idx():
+                continue
+
+            index += 1
+            task.set_last_history_idx(index)
+            data["role"] = item["role"]
+            data["thought"] = item["thought"]
+            data["content"] = item["content"]
+            yield json.dumps(data, indent = 4)
+            time.sleep(1)
```
Comment (thesues) [inline server/april/issue.py:203]: should re read file for increment swe-agent dump

File: server/april/issue.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+from logger import init_logger
+logger = init_logger(__name__)
+
+# All the project related files, include repository, files generated by agent
+# are saved in a folder generated by owner&repo under WORK_SPACE
+WORK_SPACE = "/data00/ailint/workspace"
+
+# Dict for all the tasks 
+tasks = {}
```
Comment (thesues) [inline server/april/issue.py:24]: not threas safe

File: server/april/issue.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    # Check whether there are new history data
+
+    # Send new history data back
+    index = 0
+    for item in parsed_json["history"]:
+        if item["role"] == "assistant":
+            if index < task.get_last_history_idx():
+                continue
+
+            index += 1
```
Comment (thesues) [inline server/april/issue.py:197]: if task is shared by two clients, race condition happens

File: server/april/api.py [hunk 6]
```diff
@@ -48,7 +50,42 @@ async def supported_topics() -> Response:
...
+
+        # Check if the request is for a project based prompt
+        if "prompt" in request_dict:
+            return handle_prompt(request_dict)
+
+        return JSONResponse(status_code=400, content={'message': f"The request must contain prompt"})
+    except Exception as e:
+        return JSONResponse(status_code=400, content={'message': f"invalid request, missing {e}"})
+
+@app.get("/dev/tasks/{taskId}/status", dependencies=[Depends(veriy_header)])
```
Comment (thesues) [inline server/april/api.py:79]: should be '/dev/tasks/{taskId}'

File: server/april/issue.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    if len(tokens[i]) == 0:
+        i += 1
+    return [tokens[i], tokens[i+1]]
+
+
+# The function generate the github API url, like
+#   https://api.github.com/repos/{owner}/{repo}
+#  NOTE: The function is not used for now, but keep it
+#  in case we call github api in the future
+def generateGithubApiUrl(urlObj):
```
Comment (thesues) [inline server/april/issue.py:91]: please follow PEP8 coding.

File: server/april/issue.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    promptFileName = repoFolder + "/prompt" + folderSuffix 
+    with open(promptFileName, "w") as promptFile:
+        promptFile.write(prompt)
+ 
+    newTask.set_status(TaskStatus.RUNNING)
+    # Call agent to query
+    agent = SWEAgent({"task_id": newTask.id, "data_path": promptFileName, "repo_path": repoFolder})
+    agentResp = agent.call()
+
+    return agentResp
```
Comment (thesues) [inline server/april/issue.py:167]: should return uuid

File: server/april/issue.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+import requests
+import json
+import re
+from git import Repo
+from fastapi.responses import JSONResponse, Response, StreamingResponse
+from urllib.parse import urlparse
+from github import Github
+from task import Task, TaskStatus
+from datetime import datetime
+import httpx
```
Comment (thesues) [inline server/april/issue.py:13]: unused import

File: server/april/issue.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+        i += 1
+    return [tokens[i], tokens[i+1]]
+
+
+# The function generate the github API url, like
+#   https://api.github.com/repos/{owner}/{repo}
+#  NOTE: The function is not used for now, but keep it
+#  in case we call github api in the future
+def generateGithubApiUrl(urlObj):
+    gitHubApi = urlObj.hostname
```
Comment (thesues) [inline server/april/issue.py:92]: gitlab should be supported as well

File: server/april/issue.py [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+    return [True, folderPath]
+
+# The function retrieve the owner and repository parts from url object
+def getOwnerAndRepo(urlObj)->tuple[str, str]:
+    path = urlObj.path
+    tokens = path.split("/")
+    i = 0
+    if len(tokens[i]) == 0:
+        i += 1
+    return [tokens[i], tokens[i+1]]
```
Comment (thesues) [inline server/april/issue.py:84]: [tokens[i], tokens[i+1]] is a list, not a tuple

File: server/april/issue.py [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+        return JSONResponse(status_code=200, content={'message': "SWEAgent not connected yet", "task_id": self.task_id})
+
+def generateaFolderSuffix():
+    return datetime.now().strftime("%m-%d-%Y_%H-%M-%S")
+
+# The function get the path where the repository is cloned
+# the path is created if not exists
+def getRepoFolder(owner: str, repo: str, suffix="")->tuple[bool, str]:
+    # Specify the path where you want to create the folder
+    folderPath = WORK_SPACE + f'/{owner}/{repo}'
```
Comment (thesues) [inline server/april/issue.py:53]: prefer os.path.join

File: server/april/issue.py [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+            data["role"] = item["role"]
+            data["thought"] = item["thought"]
+            data["content"] = item["content"]
+            yield json.dumps(data, indent = 4)
+            
+            # If the action contain "submit", finished
+            action = item["action"]
+            if action is not None and "submit" in action:
+                logger.info(f'Finished processing the history for task {task.get_id()}')
+                return json.dumps(empty)
```
Comment (thesues) [inline server/april/issue.py:226]: maybe jus return, instead of return json.dumpes?

File: server/april/websocket.py [hunk 14]
```diff
@@ -0,0 +1,20 @@
+
```
Comment (thesues) [inline server/april/websocket.py:1]: this file is not necessary

File: server/april/task.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+import uuid
+from datetime import datetime
+import os
+from threading import Thread, Lock
+
+class TaskStatus(Enum):
+    NEW = 0
+    INITIALIZED = 1
+    RUNNING = 2
+    DONE = 2
```
Comment (thesues) [inline server/april/task.py:11]:
```    
RUNNING = 2
DONE = 2
```
why they are the same?

File: server/april/issue.py [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+
+from logger import init_logger
+logger = init_logger(__name__)
+
+# All the project related files, include repository, files generated by agent
+# are saved in a folder generated by owner&repo under WORK_SPACE
+WORK_SPACE = "/tmp/ailint/workspace"
+
+# Dict for all the tasks 
+from threading import Thread, Lock
```
Comment (thesues) [inline server/april/issue.py:24]: Thread is unused

File: server/april/issue.py [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+    new_task = Task(prompt_obj["prompt"])
+    add_task(new_task)
+
+    folder_suffix = gen_folder_suffix()
+
+    # Clone the project to local
+    owner, repo = get_owner_repo(url_obj)
+    repo_exist, repo_folder = get_repo_folder(owner, repo, folder_suffix)
+    if not repo_exist:
+        Repo.clone_from(prompt_obj["repo"], repo_folder)
```
Comment (thesues) [inline server/april/issue.py:167]: what if this is a gitlab repo?

File: server/april/issue.py [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+                        
+            data["role"] = item["role"]
+            data["thought"] = item["thought"]
+            data["content"] = item["content"]
+            yield json.dumps(data, indent = 4)
+            
+            # If the action contain "submit", finished
+            action = item["action"]
+            if action is not None and "submit" in action:
+                logger.info(f'Finished processing the history for task {task.get_id()}')
```
Comment (thesues) [inline server/april/issue.py:225]: should call set_status....

File: server/april/task.py [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+    ERROR = -1
```
Comment (thesues) [inline server/april/task.py:12]: may this link can help you: https://fastapi.tiangolo.com/tutorial/path-params/#create-an-enum-class

File: server/april/task.py [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+    def get_last_history_idx(self):
+        with self._lock:
+            return self.last_history_idx
+    
+    def set_last_history_idx(self, index):
+        with self._lock:
+            self.last_history_idx = index
+            self.last_modified = datetime.now()
+
+    def get_last_modified(self):
```
Comment (thesues) [inline server/april/task.py:57]: who will use last_modified?

File: server/april/issue.py [hunk 21]
```diff
@@ -0,0 +1,234 @@
...
+                return json.dumps(empty)
+            
+        if index == last_index:
+            # Check whether no more histories for some period of time
+            delta = datetime.now() - task.get_last_modified
```
Comment (thesues) [inline server/april/issue.py:232]: if time.since(task.get_last_modified()). you can not delete a function.

File: server/april/api.py [hunk 22]
```diff
@@ -11,6 +11,8 @@
 from fastapi import Depends, FastAPI, HTTPException, status
 from fastapi.security import HTTPBasic, HTTPBasicCredentials
 import os
+from issue import handle_prompt, handle_status, gen_history_data
+from websocket import WebSocketServer
```
Comment (thesues) [inline server/april/api.py:15]: no websocket

File: server/april/issue.py [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+sys.path.append(swe_dir)
+
+import run
+
+from logger import init_logger
+logger = init_logger(__name__)
+
+# All the project related files, include repository, files generated by agent
+# are saved in a folder generated by owner&repo under WORK_SPACE
+WORK_SPACE = "/data00/ailint/workspace"
```
Comment (thesues) [inline server/april/issue.py:28]: use temp dir

**Pull request closed without merging**